Спринт 16 MVVM

john-krasinski opened the pull request.

Commit:
Спринт 16: Рефакторинг структуры модулей; Реализован шаблон MVVM для SearchActivity

Commit:
Спринт 16: MVVM для SettingsActivity

Т.к. по своей сути настройки приложения тесно связаны с классом App, во ViewModel оставлена логика взаимодействия с настройками через него.

Commit:
Спринт 16: MVVM для AudioPlayerActivity

kekulta commented on app/build.gradle.kts:
```diff
@@ -33,6 +33,9 @@ android {
     kotlinOptions {
         jvmTarget = "1.8"
     }
+    buildFeatures {
+        viewBinding = true
+    }
```
Молодец, что переходишь на DataBinding!

kekulta commented on app/src/main/java/com/example/playlistmaker/search/ui/SearchActivity.kt:
```diff
@@ -193,75 +203,62 @@ class SearchActivity : AppCompatActivity() {
-                reDrawHistory(tracks)
-                recyclerView.visibility = View.VISIBLE
-            }
+//            if (historyTracks.isNotEmpty()) {
```
Закоментированного кода в ПР быть не должно.
Т.к. ты используешь гит, то все предыдущие версии файлов сохранены там. Если ты удалил что-то и хочешь вернуть, то это всегда можно достать оттуда.
Закоментированный код замусоривает кодобаза и делает её менее читабельной. В больших командах так же очень просто потерять ответсвенного, который понимает нужен этот код ещё или уже нет.

kekulta commented on app/src/main/java/com/example/playlistmaker/search/ui/SearchActivity.kt:
```diff
@@ -193,75 +203,62 @@ class SearchActivity : AppCompatActivity() {
             ), onClick)
         )
-        recyclerView.visibility = View.VISIBLE
+        ui.searchRecycler.visibility = View.VISIBLE
```
Для того, чтобы писать чуть меньще кода и он выглядел слегка лаконичнее можно написать экстеншены "View.show()", "View.gone()", "View.hide()". Их можно объявить в глобальном скоупе над View, т.к. они актуально для каждого инстанса View и это часто используемые действия.

Так же, как альтернатива, можешь использовать уже существующую extension property `isVisible`.
Значение `isVisible == true` будет равно `Visibility.VISIBLE` и `isVisible == false` будет равно `Visibility.GONE`.

kekulta commented on app/src/main/java/com/example/playlistmaker/search/domain/models/Track.kt:
```diff
@@ -1,8 +1,8 @@
 
 import android.icu.text.SimpleDateFormat
-import com.example.playlistmaker.data.dto.LocalHistoryTrackDto
-import com.example.playlistmaker.data.dto.ResponseTrackDto
+import com.example.playlistmaker.search.data.dto.LocalHistoryTrackDto
+import com.example.playlistmaker.search.data.dto.ResponseTrackDto
```
Классы из пакеты data и из android. Такого не должно быть в слое domain.

kekulta commented on app/src/main/java/com/example/playlistmaker/search/ui/SearchActivity.kt:
```diff
@@ -11,64 +11,51 @@ import android.view.View
-class SearchActivity : AppCompatActivity() {
+class SearchActivity : ComponentActivity() {
+
+    private lateinit var tracksViewModel: TracksViewModel
```
Здесь лучше избегать lateinit. Так же существует делегат `by viewModel()`

Commit:
Правки в Settings

- применяется тема (возврат от ComponentActiviry к AppCompatActivity)
- фикс кнопок (Intent не запускался от AppContext, нужен был ActivityContext)

Commit:
Правки в Settings.

- применяется тема (возврат от ComponentActiviry к AppCompatActivity)
- фикс кнопок (Intent не запускался от AppContext, нужен был ActivityContext)

Commit:
Фикс границ статусбара

Commit:
Фикс отступа между списком треков и строкой поиска/заголовком, статусбара в основном экране

Commit:
Правки

- фикс отступа между списком найденных треков и полем поиска
- видимость элементов через isVisible
- создание ViewModel через by viewModels

Commit:
Чистка комментов

john-krasinski commented:
Не нашел, как поменять цвет шрифта статусбара, просто сделал статусбар светлым

Commit:
Кнопка "очистить историю" сразу после списка треков

Commit:
Белый текст статусбара при создании MainActivity
